### DISCUSSION TITLE
Why load dynamic libraries instead of linking static libraries?

### DISCUSSION THREAD

**Discussion by xuxiaocheng0201** [fzyzcjy/flutter_rust_bridge#2403] [General]
Linking static libraries can maximize the security of core logic in rust. When using a dynamic link library, others can deduce the logic of the call based on the parameters and order of the calling function.

In the output products on platforms windows and android, the figures of DLL and SO can be directly seen.

Is it because static linking is not possible on these platforms that dynamic linking is used?

**Comment by fzyzcjy** [comment]
Looks reasonable! IIRC it is because (1) Flutter's default templates use dynamic loading (2) dynamic is more flexible (3) when using static linking you should be careful not to have conflicting symbols. But if you do not have these scenarios, then yes feel free to use static!

**Reply by xuxiaocheng0201** [reply]
Yes, I don't have these scenarios. But how to use static?

**Reply by fzyzcjy** [reply]
https://cjycode.com/flutter_rust_bridge/guides/how-to/load-library

**Reply by xuxiaocheng0201** [reply]
Thank you! But I'm new to flutter, is there an example to link static library and do such load?

**Reply by fzyzcjy** [reply]
Some approaches in https://cjycode.com/flutter_rust_bridge/manual/integrate/overview may be interesting. For example, https://cjycode.com/flutter_rust_bridge/manual/integrate/template.

(They are written when frb is 1.x, so you may need modifications. Instead, maybe just look at those doc as a reference of how to create a static lib.)

**Reply by fzyzcjy** [reply]
In addition, I think your suggestion looks reasonable (static library for better security), so another potential way is to ask in Cargokit about supporting this feature.

**Reply by xuxiaocheng0201** [reply]
I saw `librust_lib_wlist_ui.a` and `rust_lib_wlist_ui.lib` in the build output directory. But I don't known how to use them.

I am not familiar with Cargokit. But I hope frb can do this thing, or provide a easy way to switch.

**Reply by fzyzcjy** [reply]
> I am not familiar with Cargokit. But I hope frb can do this thing, or provide a easy way to switch.

frb is using cargokit under the hood for the compilation (and you can customize cargokit compilation https://cjycode.com/flutter_rust_bridge/guides/how-to/rust-compilation).

**Reply by chenghang6** [reply]
It might be trivial in cargokit side.
cargokit.cmake
```
function(apply_cargokit target manifest_dir lib_name any_symbol_name)

    set(CARGOKIT_LIB_FULL_NAME "${CMAKE_SHARED_MODULE_PREFIX}${CARGOKIT_LIB_NAME}${CMAKE_SHARED_LIBRARY_SUFFIX}")
```
change it as below
```
function(apply_cargokit target manifest_dir lib_name any_symbol_name)

    set(CARGOKIT_LIB_FULL_NAME "${CMAKE_STATIC_MODULE_PREFIX}${CARGOKIT_LIB_NAME}.lib")
    message("CARGOKIT_LIB_FULL_NAME is: ${CARGOKIT_LIB_FULL_NAME}")
```

The tough part is to change ffi dynamic loading to static link/binding.

**Reply by fzyzcjy** [reply]
> The tough part is to change ffi dynamic loading to static link/binding.

https://cjycode.com/flutter_rust_bridge/guides/how-to/load-library Looks like only a few lines

**Reply by chenghang6** [reply]
I can't figure it out from your link.

**Reply by xuxiaocheng0201** [reply]
@chenghang6 Try this line:
```rust
await RustLib.init(externalLibrary: ExternalLibrary.process(iKnowHowToUseIt: true));
```
But actually I don't know how to use it exactly.

**Reply by fzyzcjy** [reply]
> But actually I don't know how to use it exactly.

Check this: https://github.com/fzyzcjy/flutter_rust_bridge/blob/ae970bfafdf80b9eb283a2167b972fb2e6504511/frb_dart/lib/src/platform_types/_io.dart#L46-L61

**Reply by xuxiaocheng0201** [reply]
Based on my understanding, is it only necessary to ensure that only a unique version of the frb library is used to use this function?

**Reply by fzyzcjy** [reply]
Hmm, IIRC some initialization function may be wrongly called, so maybe no. But if your case needs multi packages using frb, feel free to discuss!

**Reply by xuxiaocheng0201** [reply]
Can this be used in production? I tested it briefly and didn't seem to find any problems.

**Reply by chenghang6** [reply]
> Can this be used in production? I tested it briefly and didn't seem to find any problems.

Did you test it on Windows?
How did you fix this problem which was discussed before: https://github.com/fzyzcjy/flutter_rust_bridge/issues/2292?

**Reply by fzyzcjy** [reply]
Firstly make sure it is tested on a platform using static libraries. If it works in your thorough tests, then I guess it should be OK. But this is not officially supported (but feel free to PR to support it), so is not guaranteed to work forever. Instead, it would be great to combine your flutter packages that uses frb into a single one to avoid this.

**Reply by xuxiaocheng0201** [reply]
> Did you test it on Windows?

Not before, I was on macos. But I tried today on windows, it failed to build.

A temporary solution is to remove three lines in `...\build\windows\x64\cmake_install.cmake` at the appropriate time:
`".../build/windows/x64/plugins/rust_lib_wlist_ui/../rust_lib_wlist_ui"`

> How did you fix this problem which was discussed before: https://github.com/fzyzcjy/flutter_rust_bridge/issues/2292?

No, I just use frb as a normal library.

**Reply by chenghang6** [reply]
> > How did you fix this problem which was discussed before: #2292?
> 
> No, I just use frb as a normal library.

??? I thought you had it static linking.
